rules/consistent-destructuring.js

```diff
@@ -0,0 +1,140 @@
+'use strict';
+const avoidCapture = require('./utils/avoid-capture');
+const getDocumentationUrl = require('./utils/get-documentation-url');
+
+const MESSAGE_ID = 'consistentDestructuring';
+const MESSAGE_ID_SUGGEST = 'consistentDestructuringSuggest';
+
+const declaratorSelector = [
+	'VariableDeclarator',
+	'[id.type="ObjectPattern"]',
+	'[init.type!="Literal"]'
+].join('');
+
+const isChildInParentScope = (child, parent) => {
+	while (child) {
+		if (child === parent) {
+			return true;
+		}
+
+		child = child.upper;
+	}
+
+	return false;
+};
+
+const fixDestructuring = (fixer, objectPattern, member, newMember) => {
+	// Check if member needs to be renamed
+	const property = member === newMember ?
+		`${member}` :
+		`${member}: ${newMember}`;
+	const {properties} = objectPattern;
+	const lastProperty = properties[properties.length - 1];
+
+	if (lastProperty) {
+		return fixer.insertTextAfter(lastProperty, `, ${property}`);
+	}
+
+	return fixer.replaceText(objectPattern, `{${property}}`);
+};
+
+const create = context => {
+	const {ecmaVersion} = context.parserOptions;
+	const source = context.getSourceCode();
+	const declarations = new Map();
+
+	return {
+		[declaratorSelector]: node => {
+			declarations.set(source.getText(node.init), {
+				scope: context.getScope(),
+				variables: context.getDeclaredVariables(node),
+				objectPattern: node.id
+			});
+		},
+		MemberExpression: node => {
+			const {parent, object, property} = node;
+
+			// Ignore member function calls and delete expressions
+			if (
+				(parent.type === 'CallExpression' &&
+				parent.callee === node) ||
+				(parent.type === 'UnaryExpression' &&
+				parent.operator === 'delete')
+			) {
+				return;
+			}
```

Comment:
```js
MemberExpression
:not(
	CallExpression > MemberExpression.callee,
	UnaryExpression[operator="delete"] > MemberExpression.argument,
)
```

Want try this selector? It can do the check